src/app/data-client.ts

```diff
@@ -0,0 +1,58 @@
+import { type RpcOptions } from '@improbable-eng/grpc-web/dist/typings/client.d';
+import {
+  DataRequest,
+  Filter,
+  TabularData,
+  TabularDataByFilterRequest,
+  TabularDataByFilterResponse,
+} from '../gen/app/data/v1/data_pb';
+import { DataServiceClient } from '../gen/app/data/v1/data_pb_service';
+import { promisify } from '../utils';
+
+export class DataClient {
+  private client: DataServiceClient;
+
+  constructor(serviceHost: string, grpcOptions: RpcOptions) {
+    this.client = new DataServiceClient(serviceHost, grpcOptions);
+  }
+
+  private get service() {
+    return this.client;
+  }
+
+  async tabularDataByFilter(filter: Filter | undefined) {
+    const { service } = this;
+
+    let last = '';
+    const dataArray: TabularData.AsObject[] = [];
+
+    for (;;) {
+      const dataReq = new DataRequest();
+      if (filter) {
+        dataReq.setFilter(filter);
+      } else {
+        dataReq.setFilter(new Filter());
```

Comment:
how hard is it to create a new filter? should we have convenience methods like in the python sdk?

Reply:
It's not that hard to create a filter (there's an example of it in the example). It doesn't seem necessary to have such a method.

Reply:
how does one convert a Date object into a timestamp?

Reply:
The Timestamp object has a `fromDate()` function.